### ISSUE TITLE
Error applying mean node on ImageInput (Examples/Image/MNIST/02_Convolution.ndl)

### ISSUE THREAD

**Issue by tz579** [Microsoft/CNTK#181]
Regarding to Examples/Image/MNIST/02_Convolution.ndl, if the following lines are added below the "ImageInput" line:

    features = ImageInput(imageW, imageH, 1, imageLayout=$imageLayout$)  #Original Line
    featMean = Mean(features)  #Line Added
    featStdDev = InvStdDev(features)  #Line Added

There would be dimensionality error during **PreCompute** stage, shown as below:

    [CALL STACK]
      /usr/local/cntk/lib/libcntkmath.so ( Microsoft::MSR::CNTK::ExceptionWithCallStack<std::logic_error>::CollectCallStack(std::function<void (std::string)> const&, std::function<void ()> const&)  + 0x101 )  [0x7f525c1476f1]
      /usr/local/cntk/lib/libcntkmath.so ( Microsoft::MSR::CNTK::ExceptionWithCallStack<std::logic_error>::PrintCallStack()  + 0x88  )  [0x7f525c147b88]
      cntk                           ( void Microsoft::MSR::CNTK::ThrowFormatted<std::logic_error>(char const*, ...)  + 0xe2  )  [0x543ae2]
      cntk                           ( Microsoft::MSR::CNTK::ComputationNode<float>::BeginForwardProp()  + 0x215 )  [0x546865]
      cntk                           ( Microsoft::MSR::CNTK::ComputationNetwork::PARTraversalFlowControlNode::ForwardProp(Microsoft::MSR::CNTK::FrameRange const&)  + 0x56  )  [0x6b25e6]
      cntk                           ( Microsoft::MSR::CNTK::ComputationNetwork::ForwardProp(std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>)  + 0x7c  )  [0x6b4a2c]
      cntk                           ( Microsoft::MSR::CNTK::SGD<float>::PreCompute(std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork>, Microsoft::MSR::CNTK::IDataReader<float>*, std::vector<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>, std::allocator<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase> > >&, std::vector<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>, std::allocator<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase> > >&, std::map<std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, Microsoft::MSR::CNTK::Matrix<float>*, std::less<std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > >, std::allocator<std::pair<std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const, Microsoft::MSR::CNTK::Matrix<float>*> > >*)  + 0x2ec )  [0x72074c]
      cntk                           ( Microsoft::MSR::CNTK::SGD<float>::TrainOrAdaptModel(int, std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork>, bool, std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork>, std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>, Microsoft::MSR::CNTK::IDataReader<float>*, Microsoft::MSR::CNTK::IDataReader<float>*)  + 0x9a4 )  [0x729be4]
      cntk                           ( Microsoft::MSR::CNTK::SGD<float>::Train(std::function<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork> (int)>, int, Microsoft::MSR::CNTK::IDataReader<float>*, Microsoft::MSR::CNTK::IDataReader<float>*, bool)  + 0x424 )  [0x72c2a4]
      cntk                           ( void DoTrain<Microsoft::MSR::CNTK::ConfigParameters, float>(Microsoft::MSR::CNTK::ConfigParameters const&)  + 0x22c )  [0x74804c]
      cntk                           ( void DoCommands<float>(Microsoft::MSR::CNTK::ConfigParameters const&)  + 0x804 )  [0x596cd4]
      cntk                           ( wmainOldCNTKConfig(int, wchar_t**)        + 0x909 )  [0x53abe9]
      cntk                           ( wmain1(int, wchar_t**)                    + 0x70  )  [0x53b240]
      cntk                           ( main                                      + 0xd8  )  [0x52fec8]
      /lib/x86_64-linux-gnu/libc.so.6 ( __libc_start_main                         + 0xf0  )  [0x7f525ad15a40]
      cntk                           ( _start                                    + 0x29  )  [0x533c99]
    EXCEPTION occurred: VerifySize: expected matrix size 28 x 28, but it is 784 x 1

This happens **only for the Mean node but not for the InvStdDev node**, and only after **commit 14fee02 on Jan 21, 2016** (I tracked it back and found that commit 14fee02 led to this "bug", but it is related to several commits before and after that).

**Comment by frankseide** [comment]
This is a bug in memory sharing. A fix is in branch fseide/s2s. Still waiting for some CRs before I can integrate with master.

**Comment by frankseide** [comment]
I think this was the fix: Call MarkValueNonSharable() in the PrecomputeNode constructor. I vaguely remember that this fixed it, although we made more changes. You can give it a try or wait for the merge.

    PreComputedNodeBase(DEVICEID_TYPE deviceId, const wstring& name)
        : Base(deviceId, name), m_hasComputed(false)
    {
        MarkValueNonSharable();
    }

**Comment by tz579** [comment]
Thanks a lot! I have just pulled the branch fseide/s2s and tested it (**using the *.ndl *.cntk files in this branch, adding the same Mean and InvStdDev nodes as mentioned above**). Unfortunately the problem persists. The error information changes a little bit though:
    EXCEPTION occurred: VerifySize: expected matrix size 28 x 28, but it is 784 x 1, for featMean Mean operation.
    
    [CALL STACK]
      cntk                           ( Microsoft::MSR::CNTK::ComputationNode<float>::BeginForwardProp()  + 0x216 )  [0x570a96]
      cntk                           ( Microsoft::MSR::CNTK::ComputationNetwork::PARTraversalFlowControlNode::ForwardProp(Microsoft::MSR::CNTK::FrameRange const&)  + 0x56  )  [0x6bf1d6]
      cntk                           ( Microsoft::MSR::CNTK::ComputationNetwork::ForwardProp(std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>)  + 0x7c  )  [0x6c161c]
      cntk                           ( Microsoft::MSR::CNTK::SGD<float>::PreCompute(std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork>, Microsoft::MSR::CNTK::IDataReader*, std::vector<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>, std::allocator<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase> > >&, std::vector<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>, std::allocator<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase> > >&, Microsoft::MSR::CNTK::StreamMinibatchInputs*)  + 0x304 )  [0x73f9b4]
      cntk                           ( Microsoft::MSR::CNTK::SGD<float>::TrainOrAdaptModel(int, std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork>, bool, std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork>, std::shared_ptr<Microsoft::MSR::CNTK::ComputationNodeBase>, Microsoft::MSR::CNTK::IDataReader*, Microsoft::MSR::CNTK::IDataReader*)  + 0x9b7 )  [0x745977]
      cntk                           ( Microsoft::MSR::CNTK::SGD<float>::Train(std::function<std::shared_ptr<Microsoft::MSR::CNTK::ComputationNetwork> (int)>, int, Microsoft::MSR::CNTK::IDataReader*, Microsoft::MSR::CNTK::IDataReader*, bool)  + 0x424 )  [0x748074]
      cntk                           ( void DoTrain<Microsoft::MSR::CNTK::ConfigParameters, float>(Microsoft::MSR::CNTK::ConfigParameters const&)  + 0x22c )  [0x755cdc]
      cntk                           ( void DoCommands<float>(Microsoft::MSR::CNTK::ConfigParameters const&)  + 0x8d9 )  [0x59dd49]
      cntk                           ( wmainOldCNTKConfig(int, wchar_t**)        + 0x8ab )  [0x54017b]
      cntk                           ( wmain1(int, wchar_t**)                    + 0x68  )  [0x5407b8]
      cntk                           ( main                                      + 0xd8  )  [0x534aa8]
      /lib/x86_64-linux-gnu/libc.so.6 ( __libc_start_main                         + 0xf0  )  [0x7f3bd8366a40]
      cntk                           ( _start                                    + 0x29  )  [0x538829]

**Comment by frankseide** [comment]
I now realize that the dimensions match, 28 * 28 = 784. So this is a broken tensor dimension.

Could you please paste the Validate() output? This will show the dimensions of all elements.

**Comment by tz579** [comment]
Validating network. 30 nodes to process in pass 1.
    
    Validating --> labels = InputValue -> [10 x *]
    Validating --> ol.W = LearnableParameter -> [10 x 128]
    Validating --> h1.W = LearnableParameter -> [128 x 4 x 4 x 32]
    Validating --> conv2_act.convW = LearnableParameter -> [32 x 400]
    Validating --> conv1_act.convW = LearnableParameter -> [16 x 25]
    Validating --> featScale = LearnableParameter -> [1 x 1]
    Validating --> features = InputValue -> [28 x 28 x 1 x *]
    Validating --> featScaled = ElementTimes(featScale[1 x 1], features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1 x *]
    Validating --> conv1_act.conv = Convolution(conv1_act.convW[16 x 25], featScaled[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [24 x 24 x 16 x *]
    Validating --> conv1_act.convB = LearnableParameter -> [1 x 1 x 16]
    Validating --> conv1_act.convPlusB = Plus(conv1_act.conv[24 x 24 x 16 x * {W=24, H=16, C=24}], conv1_act.convB[1 x 1 x 16]) -> [24 x 24 x 16 x *]
    Validating --> conv1_act.act = RectifiedLinear(conv1_act.convPlusB[24 x 24 x 16 x * {W=24, H=16, C=24}]) -> [24 x 24 x 16 x *]
    Validating --> pool1 = MaxPooling(conv1_act.act[24 x 24 x 16 x * {W=24, H=16, C=24}]) -> [12 x 12 x 16 x *]
    Validating --> conv2_act.conv = Convolution(conv2_act.convW[32 x 400], pool1[12 x 12 x 16 x * {W=12, H=16, C=12}]) -> [8 x 8 x 32 x *]
    Validating --> conv2_act.convB = LearnableParameter -> [1 x 1 x 32]
    Validating --> conv2_act.convPlusB = Plus(conv2_act.conv[8 x 8 x 32 x * {W=8, H=32, C=8}], conv2_act.convB[1 x 1 x 32]) -> [8 x 8 x 32 x *]
    Validating --> conv2_act.act = RectifiedLinear(conv2_act.convPlusB[8 x 8 x 32 x * {W=8, H=32, C=8}]) -> [8 x 8 x 32 x *]
    Validating --> pool2 = MaxPooling(conv2_act.act[8 x 8 x 32 x * {W=8, H=32, C=8}]) -> [4 x 4 x 32 x *]
    Validating --> h1.t = Times(h1.W[128 x 4 x 4 x 32], pool2[4 x 4 x 32 x * {W=4, H=32, C=4}]) -> [128 x *]
    Validating --> h1.b = LearnableParameter -> [128 x 1]
    Validating --> h1.z = Plus(h1.t[128 x *], h1.b[128 x 1]) -> [128 x 1 x *]
    Validating --> h1.y = Sigmoid(h1.z[128 x 1 x *]) -> [128 x 1 x *]
    Validating --> ol.t = Times(ol.W[10 x 128], h1.y[128 x 1 x *]) -> [10 x 1 x *]
    Validating --> ol.b = LearnableParameter -> [10 x 1]
    Validating --> ol.z = Plus(ol.t[10 x 1 x *], ol.b[10 x 1]) -> [10 x 1 x *]
    Validating --> ce = CrossEntropyWithSoftmax(labels[10 x *], ol.z[10 x 1 x *]) -> [1]
    Validating --> err = ErrorPrediction(labels[10 x *], ol.z[10 x 1 x *]) -> [1]
    Validating --> featMean = Mean(features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1]
    Validating --> featStdDev = InvStdDev(features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1]
    Validating --> featNorm = PerDimMeanVarNormalization(features[28 x 28 x 1 x * {W=28, H=1, C=28}], featMean[28 x 28 x 1 {W=28, H=1, C=28}], featStdDev[28 x 28 x 1 {W=28, H=1, C=28}]) -> [28 x 28 x 1 x *]
    
    Validating network. 19 nodes to process in pass 2.
    
    Validating --> labels = InputValue -> [10 x *]
    Validating --> ol.W = LearnableParameter -> [10 x 128]
    Validating --> h1.W = LearnableParameter -> [128 x 4 x 4 x 32]
    Validating --> conv2_act.convW = LearnableParameter -> [32 x 400]
    Validating --> conv1_act.convW = LearnableParameter -> [16 x 25]
    Validating --> featScale = LearnableParameter -> [1 x 1]
    Validating --> features = InputValue -> [28 x 28 x 1 x *]
    Validating --> featScaled = ElementTimes(featScale[1 x 1], features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1 x *]
    Validating --> conv1_act.conv = Convolution(conv1_act.convW[16 x 25], featScaled[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [24 x 24 x 16 x *]
    Validating --> conv1_act.convB = LearnableParameter -> [1 x 1 x 16]
    Validating --> conv1_act.convPlusB = Plus(conv1_act.conv[24 x 24 x 16 x * {W=24, H=16, C=24}], conv1_act.convB[1 x 1 x 16]) -> [24 x 24 x 16 x *]
    Validating --> conv1_act.act = RectifiedLinear(conv1_act.convPlusB[24 x 24 x 16 x * {W=24, H=16, C=24}]) -> [24 x 24 x 16 x *]
    Validating --> pool1 = MaxPooling(conv1_act.act[24 x 24 x 16 x * {W=24, H=16, C=24}]) -> [12 x 12 x 16 x *]
    Validating --> conv2_act.conv = Convolution(conv2_act.convW[32 x 400], pool1[12 x 12 x 16 x * {W=12, H=16, C=12}]) -> [8 x 8 x 32 x *]
    Validating --> conv2_act.convB = LearnableParameter -> [1 x 1 x 32]
    Validating --> conv2_act.convPlusB = Plus(conv2_act.conv[8 x 8 x 32 x * {W=8, H=32, C=8}], conv2_act.convB[1 x 1 x 32]) -> [8 x 8 x 32 x *]
    Validating --> conv2_act.act = RectifiedLinear(conv2_act.convPlusB[8 x 8 x 32 x * {W=8, H=32, C=8}]) -> [8 x 8 x 32 x *]
    Validating --> pool2 = MaxPooling(conv2_act.act[8 x 8 x 32 x * {W=8, H=32, C=8}]) -> [4 x 4 x 32 x *]
    Validating --> h1.t = Times(h1.W[128 x 4 x 4 x 32], pool2[4 x 4 x 32 x * {W=4, H=32, C=4}]) -> [128 x *]
    Validating --> h1.b = LearnableParameter -> [128 x 1]
    Validating --> h1.z = Plus(h1.t[128 x *], h1.b[128 x 1]) -> [128 x 1 x *]
    Validating --> h1.y = Sigmoid(h1.z[128 x 1 x *]) -> [128 x 1 x *]
    Validating --> ol.t = Times(ol.W[10 x 128], h1.y[128 x 1 x *]) -> [10 x 1 x *]
    Validating --> ol.b = LearnableParameter -> [10 x 1]
    Validating --> ol.z = Plus(ol.t[10 x 1 x *], ol.b[10 x 1]) -> [10 x 1 x *]
    Validating --> ce = CrossEntropyWithSoftmax(labels[10 x *], ol.z[10 x 1 x *]) -> [1]
    Validating --> err = ErrorPrediction(labels[10 x *], ol.z[10 x 1 x *]) -> [1]
    Validating --> featMean = Mean(features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1]
    Validating --> featStdDev = InvStdDev(features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1]
    Validating --> featNorm = PerDimMeanVarNormalization(features[28 x 28 x 1 x * {W=28, H=1, C=28}], featMean[28 x 28 x 1 {W=28, H=1, C=28}], featStdDev[28 x 28 x 1 {W=28, H=1, C=28}]) -> [28 x 28 x 1 x *]
    
    Validating network, final pass.
    
    Validating --> labels = InputValue -> [10 x *]
    Validating --> ol.W = LearnableParameter -> [10 x 128]
    Validating --> h1.W = LearnableParameter -> [128 x 4 x 4 x 32]
    Validating --> conv2_act.convW = LearnableParameter -> [32 x 400]
    Validating --> conv1_act.convW = LearnableParameter -> [16 x 25]
    Validating --> featScale = LearnableParameter -> [1 x 1]
    Validating --> features = InputValue -> [28 x 28 x 1 x *]
    Validating --> featScaled = ElementTimes(featScale[1 x 1], features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1 x *]
    Validating --> conv1_act.conv = Convolution(conv1_act.convW[16 x 25], featScaled[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [24 x 24 x 16 x *]
    Validating --> conv1_act.convB = LearnableParameter -> [1 x 1 x 16]
    Validating --> conv1_act.convPlusB = Plus(conv1_act.conv[24 x 24 x 16 x * {W=24, H=16, C=24}], conv1_act.convB[1 x 1 x 16]) -> [24 x 24 x 16 x *]
    Validating --> conv1_act.act = RectifiedLinear(conv1_act.convPlusB[24 x 24 x 16 x * {W=24, H=16, C=24}]) -> [24 x 24 x 16 x *]
    Validating --> pool1 = MaxPooling(conv1_act.act[24 x 24 x 16 x * {W=24, H=16, C=24}]) -> [12 x 12 x 16 x *]
    Validating --> conv2_act.conv = Convolution(conv2_act.convW[32 x 400], pool1[12 x 12 x 16 x * {W=12, H=16, C=12}]) -> [8 x 8 x 32 x *]
    Validating --> conv2_act.convB = LearnableParameter -> [1 x 1 x 32]
    Validating --> conv2_act.convPlusB = Plus(conv2_act.conv[8 x 8 x 32 x * {W=8, H=32, C=8}], conv2_act.convB[1 x 1 x 32]) -> [8 x 8 x 32 x *]
    Validating --> conv2_act.act = RectifiedLinear(conv2_act.convPlusB[8 x 8 x 32 x * {W=8, H=32, C=8}]) -> [8 x 8 x 32 x *]
    Validating --> pool2 = MaxPooling(conv2_act.act[8 x 8 x 32 x * {W=8, H=32, C=8}]) -> [4 x 4 x 32 x *]
    Validating --> h1.t = Times(h1.W[128 x 4 x 4 x 32], pool2[4 x 4 x 32 x * {W=4, H=32, C=4}]) -> [128 x *]
    Validating --> h1.b = LearnableParameter -> [128 x 1]
    Validating --> h1.z = Plus(h1.t[128 x *], h1.b[128 x 1]) -> [128 x 1 x *]
    Validating --> h1.y = Sigmoid(h1.z[128 x 1 x *]) -> [128 x 1 x *]
    Validating --> ol.t = Times(ol.W[10 x 128], h1.y[128 x 1 x *]) -> [10 x 1 x *]
    Validating --> ol.b = LearnableParameter -> [10 x 1]
    Validating --> ol.z = Plus(ol.t[10 x 1 x *], ol.b[10 x 1]) -> [10 x 1 x *]
    Validating --> ce = CrossEntropyWithSoftmax(labels[10 x *], ol.z[10 x 1 x *]) -> [1]
    Validating --> err = ErrorPrediction(labels[10 x *], ol.z[10 x 1 x *]) -> [1]
    Validating --> featMean = Mean(features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1]
    Validating --> featStdDev = InvStdDev(features[28 x 28 x 1 x * {W=28, H=1, C=28}]) -> [28 x 28 x 1]
    Validating --> featNorm = PerDimMeanVarNormalization(features[28 x 28 x 1 x * {W=28, H=1, C=28}], featMean[28 x 28 x 1 {W=28, H=1, C=28}], featStdDev[28 x 28 x 1 {W=28, H=1, C=28}]) -> [28 x 28 x 1 x *]
    
    13 out of 30 nodes do not share the minibatch layout with the input data.
    
    Post-processing network complete.
    
    SGD using GPU 0.
    
    Training criterion node(s):
    	ce = CrossEntropyWithSoftmax
    
    Evaluation criterion node(s):
    	err = ErrorPrediction
    
    
    Allocating matrices for forward and/or backward propagation.
    
    Precomputing --> 2 PreCompute nodes found.
    
    	NodeName: featMean
    	NodeName: featStdDev
    UCIFastReader: Starting at epoch 0, counting lines to determine record count...
     60000 records found.
    starting epoch 0 at record count 0, and file position 0
    already there from last epoch
    RandomOrdering: 11898 retries for 60000 elements (19.8%) to ensure window condition
    RandomOrdering: recached sequence for seed 0: 25151, 18078, ...
    
    EXCEPTION occurred: VerifySize: expected matrix size 28 x 28, but it is 784 x 1, for featMean Mean operation.
Thanks! Actually cntk **never** complains for the validation for any of the 3 passes. **It only complains after that.**

**Comment by frankseide** [comment]
Hmm... This should not happen. It got the correct dimensions except for the storage object. I will debug this, but probably I won't finish today.

**Comment by tz579** [comment]
Thanks a lot! Please note again that **this only happens for the Mean node added to the original NDL, not for the InvStdDev node**.

**Comment by tz579** [comment]
Thanks a lot! Please note again that this occurs **only on the Mean node added to the original NDL (as mentioned in the 1st post)**, not for the InvStdDev node, and **only after commit 14fee02 on Jan 21, 2016**.

**Comment by frankseide** [comment]
Quick update: I have a code fix, but the results differ marginally. Before releasing, we still need to verify whether that is due to summation order or an bug.

**Comment by tz579** [comment]
@frankseide Thanks a lot! After all it was weird for that bug to appear after commit 14fee02 and only for the Mean node.

**Comment by frankseide** [comment]
Actually also for InvStdDev node. Working on it.

**Comment by frankseide** [comment]
Fixed. Will land in master once code review and lab test are finished (few hours from now).

**Comment by tz579** [comment]
@frankseide Wow thanks a lot! I have tested the code on fseide/pack and it's working correctly on my side, too!

**Comment by frankseide** [comment]
Thanks so much for letting me know!

So far tests are green, so with luck, it will go live in an hour.

**Comment by frankseide** [comment]
Fix is in master.

**Comment by tz579** [comment]
@frankseide Thanks a lot! After further tests, **a new issue (#210)** has been opened to help clarify **an additional error** (may or may not come from the same bug) posterior to this fix.
